src/leveled_penciller.erl

```diff
@@ -1613,338 +1615,236 @@ compare_to_sqn(ObjSQN, _SQN) when is_integer(ObjSQN) ->
 compare_to_sqn(Obj, SQN) ->
     compare_to_sqn(leveled_codec:strip_to_seqonly(Obj), SQN).
 
+-spec maybelog_fetch_timing(
+    leveled_monitor:monitor(),
+    memory|leveled_pmanifest:lsm_level(),
+    leveled_monitor:timing(),
+    boolean()) -> ok.
+maybelog_fetch_timing(_Monitor, _Level, no_timing, _NF) ->
+    ok;
+maybelog_fetch_timing({Pid, _StatsFreq}, _Level, FetchTime, true) ->
+    leveled_monitor:add_stat(Pid, {pcl_fetch_update, not_found, FetchTime});
+maybelog_fetch_timing({Pid, _StatsFreq}, Level, FetchTime, _NF) ->
+    leveled_monitor:add_stat(Pid, {pcl_fetch_update, Level, FetchTime}).
 
 %%%============================================================================
-%%% Iterator functions 
-%%% 
-%%% TODO - move to dedicated module with extended unit testing
+%%% Key folder
 %%%============================================================================
 
+-type sst_iterator()
+    :: #{
+        leveled_pmanifest:lsm_level() =>
+            list(leveled_sst:expandable_pointer()|leveled_codec:ledger_kv()),
+        -1 =>
+            list(leveled_codec:ledger_kv())}.
+-type max_keys() :: unlimited|non_neg_integer().
+-type iterator_level() :: -1|leveled_pmanifest:lsm_level().
+-type search_info() ::
+    {{leveled_codec:ledger_key(), leveled_codec:ledger_key()},
+        {non_neg_integer(), pos_integer()|infinity},
+        leveled_sst:segment_check_fun()}.
+
+-define(NULL_KEY, {null, null}).
 
 -spec keyfolder(
-    {list(), list()},
+    sst_iterator(),
     {leveled_codec:ledger_key(), leveled_codec:ledger_key()},
     {pclacc_fun(), any(), pos_integer()},
     {leveled_sst:segment_check_fun(),
         {non_neg_integer(), pos_integer()|infinity},
-        integer()}) -> any().
+        -1|non_neg_integer()}) -> {non_neg_integer(), term()}|term().
+keyfolder(
+        Iterator,
+        {StartKey, EndKey},
+        {AccFun, InitAcc, Now},
+        {SegCheckFun, LastModRange, KeyLimit}) ->
+    % The in-memory dump of keys in this range, may go beyond the end key - so
+    % strip these back before starting the fold 
+    StripIMMFun =
+        fun(MemIter) ->
+            lists:reverse(
```

Comment:
this looks expensive, reverse twice and dropwhile.
Isn't this the same as a takewhile without reverses?

Reply:
No it is not the same. The reverse twice isn't that bad at all. I tried a few other ways and was not successful, so the penalty for list reverse in Erlang is very low. No reason to worry.
For example, a `foldr` would be more expensive.

Reply:
Agree.  I did some local tests and reached the same conclusion.